### DISCUSSION TITLE
`let`, `var`, `fn` -> `let` ?

### DISCUSSION THREAD

**Discussion by christianparpart** [carbon-language/carbon-lang#1855] [Language design]
Hey all,

thanks for the great project at first. I am having some concerns and confusion about the differenciation of `let` vs `var` sometimes. While I understand (I hope) that `var` is used for things that have an address, to me this feels like way to many different ways of declaring a symbol. This is why I took the change list `fn` to my confusion-list. What I really appreciate about F# language is that there's **only** `let` for declaring by default immutable but also mutable variables as well as (!) functions. I wonder if it might make sense to (since the language claim is to be "simple") only have one keyword for declaring symbols and then probably have a modifier to allow access to its address. 

Why could just a proposal be a bad one?

To add some imaginary example:
```fsharp
let pi: f32 = 3.1415;

let address s: auto = 1.2; // with address being the modifier, whatever word would fit best, that's debatable.

let mut changable c: u32 = 42; // some mutable variable

let f (x: f32) -> f32 {
    return s * x + 1.0;
}
```

**Comment by jonmeow** [comment]
@christianparpart I've edited your post. Please use gender-neutral language such as "everyone", "folks", or "friends" when addressing the community as a whole. Not everyone here is a "guy". Thanks!

**Reply by christianparpart** [reply]
I didn't think of it that way. Interesting, thanks.

**Comment by jonmeow** [comment]
Note, what you're proposing is already discussed on #1591

**Reply by christianparpart** [reply]
Withe respect to uniforming function syntax, yeah, you're right, even though their syntax proposal was different. I am looking at it from an F#'s language point of view and that was what I (amongst others) enjoyed the most.

I am no master of Carbon language, but maybe we can also get rid of the `var` keyword to introduce a variable along `let` and only use `let` with possibly modifiers to it if needed (for declaring it as mutable or addressible).

**Reply by jonmeow** [reply]
@christianparpart Using "its" to refer to a person is often dehumanizing, don't use it by default. When you don't know the right pronoun, "his/her" is typically fine, you can also say "their". https://english.stackexchange.com/questions/89937/why-does-it-have-a-dehumanizing-connotation

**Reply by christianparpart** [reply]
Interesting. Thanks.

**Reply by uspasojevic96** [reply]
@jonmeow so when do you plan to start talking about the actual topic here instead of being a neogrammar nazi?

**Reply by christianparpart** [reply]
Oha. I would have chosen other words. But @jonmeow, this guy has a point. I was asking kindly and tried to be polite in all the possible dimensions that I could think of at that time. Nobody's perfect, but I'd prefer us (all) to keep focus and not over-do the reinterpretation of words. I am a non-native English speaker, I apologize for not always taking the right words (giving my best though), I am legally blind, so even when searching I might not always find existing topics, especially since the one you posted was not the main issue of what I actually wanted to talk about. Now this. All this discussion and zero actual technical content makes me feel already sad, and on one part dehumanized and bullied myself just by the fact that I feel I am forced out with my personal health state. No worries, I'm used to that. Let's just assume we're all here on good intent, and I will continue to do so in the future here, too (I hope I'm still welcome). Keep up the good Carbon-work. Happy Monday to all.

**Reply by L4stR1t3s** [reply]
A more constructive criticism:

I think that there is an unreasonable expectation here of non-native English users, especially those who are not very fluent in English.

> Please use gender-neutral language ...

Non-native English users translate from their own language, and gender-neutral language is not the same in every language. The literal translation of a word may be gender neutral in one language, but not in the other. It is unreasonable to ask someone who isn't fluent in a specific language to use gender neutral language in general because they lack the ability to do so.

> When you don't know the right pronoun ...

Someone who is not skilled in English should not be asked to evaluate the correctness of their own use of the English language.

There already is a bigger barrier of entry for non-native English speakers to take part in a conversation. Translating from your own language, using translation tools if you don't know certain words, is more time consuming. There is always a worry that you might not be communicating your message well. Now someone also has to worry not to cause offense by using inappropriate language, a distinction they are unable to make. People might not take part in a conversation because they lack the knowledge to know if they are violating these rules or not.

So in order to be more inclusive to non-native English speakers, especially those who are not very fluent in English I would like to suggest the following: When correcting someone, only focus on the specific mistake they make, and refrain from using general statements. Please don't use ... because ..., use ... instead.

EDIT: Forgot to say that you should definitely not make this type of statement:

>Not everyone here is a "guy".

That may be obvious to you, but it is not obvious to someone whose primary language is not English. At least explain that "guys" refers to men only, don't assume people have the same knowledge of the English language you do.

**Reply by uspasojevic96** [reply]
@L4stR1t3s while I do appreciate you defending non-native English speakers, keep in mind that there are people like me which don't care about pronouns and other politically correct crap that people have been spewing.

**Reply by L4stR1t3s** [reply]
@uspasojevic96 If you didn't care about something, you wouldn't complain about it so much when others do care about it. I don't care about cars, don't have one, never learned to drive, but I don't go around complaining about people that do.

**Reply by jonmeow** [reply]
@L4stR1t3s Thank you for the commentary on my messaging style. I'll continue working to improve it. I've been trying to soften my corrections but it's hard to get that correctly over the internet, I know I'm not perfect. But the issues of [referring to everyone as "guys"](https://www.includecpp.org/resources/language/#guys) is important to us, particularly because it builds upon longstanding issues in tech communities being predominately men, and women and non-binary folk feeling pushed aside.

@christianparpart I'm sorry my words have hurt you; again, I need to keep improving how I moderate. However, we need to keep moderating. While I understand that this is a non-technical discussion, we do have to think about how we conduct ourselves as a community. While I've been assuming good intent on your part, it's also important to bear in mind that your words have effects on others and send a message about who is welcome. These non-technical aspects are very important in order to ensure we can all comfortably discuss technical aspects. We need to keep trying to do better.

@uspasojevic96 It's important to be welcoming to community members as part of our [code of conduct](https://github.com/carbon-language/carbon-lang/blob/trunk/CODE_OF_CONDUCT.md). We do not let those who cannot be thoughtful about others participate in community forums.

**Comment by christianparpart** [comment]
It's now quite some time ago. Apparently no change in the language with respect to that. I wonder why?

We just locally talked about the big hopes of Carbon among other devs and realized that `var` is not a good pick especially also because it's visually not to distinguishable from `let` as it's both of the same length and therefore makes it harder(!) for a code reviewer to spot it.
Simply going for `let` vs `let mut` is well proven in other languages already, and I do not see why Carbon should divert here, while already borrowing other parts.

So what is your stance on this one? Many thanks.

**Reply by chandlerc** [reply]
> It's now quite some time ago. Apparently no change in the language with respect to that. I wonder why?

So far, we haven't had a concrete proposal to change. At this point, it remains just a discussion thread AFAIK.

> We just locally talked about the big hopes of Carbon among other devs and realized that `var` is not a good pick especially also because it's visually not to distinguishable from `let` as it's both of the same length and therefore makes it harder(!) for a code reviewer to spot it. Simply going for `let` vs `let mut` is well proven in other languages already, and I do not see why Carbon should divert here, while already borrowing other parts.

Using `var` isn't without precedent. Both Java and C# use `var` in a similar introducer position, and Kotlin's usage is also very similar.

Moreover, Swift specifically uses both `var` and `let` together without issue or complaints from users.

To a certain extent, I think there a wide range of possible designs for the introducers in this space, and many of them are reasonably workable. Personally, I don't think one spelling or structure vs. another of the introducer keywords here is going to be an especially large factor one way or the other in Carbon's success. But we need to pick something, and so we did it based on our best judgement of the tradeoffs present.

For example, Kotlin using `var` and `val` which seemed *even harder* to distinguish. We felt like `var` and `let` were a good balance between usefully consistent structure/size and helpfully different shape of the letters. Plus, Swift helped test out the ground of this combination and didn't encounter any problems.

Anyways, just my personal view on these options.

> So what is your stance on this one? Many thanks.

While the above is my personal stance on the tradeoff, the issue linked earlier actually has a great summary of what needs to happen if folks feel strongly about changing direction here:
https://github.com/carbon-language/carbon-lang/issues/1591#:~:text=So%20I%20think%20to%20consider%20this%2C%20we%27d%20need%20both%20a%20signal%20of%20clear%20community%20interest%20in%20the%20direction%20and%20good%20rationale%20write%2Dup%20around%20the%20benefits%20that%20the%20uniform%20approach%20would%20take.

At the end of the day, we'll need significant community interest in changing here combined with a clear proposal and write of rationale.

**Reply by jens-auer** [reply]
The fact that var and let have the same length is an advantage for me since it makes variables and constants aligned:
```
let pi = 3.14;
var area =  pi * r * r
```
is nicely aligned if you use a monospace font.
